ament_vendor: Add AMENT_VENDOR_POLICY CMake option

traversaro opened the pull request:
Evolution of https://github.com/ament/ament_cmake/pull/552 .

This PR aim is based on our experience at RoboStack with the ament_vendor, and aims to improve the experience of downstream packagers of ROS on non-officially supported platform (being them [conda/robostack](https://robostack.github.io/), [nix](https://github.com/lopsided98/nix-ros-overlay) or similar), reducing the need of patching the vendor package to obtain the required behaviour.

In particular, a common pattern that we are experiencing is that compare to stable linux distributions, rolling releases as conda-forge have much more recent packages, so in many case we want to disable completely the vendoring in `_vendor` packages. 

A few examples:
* `zenoh_cpp_vendor`: In  conda-forge, we have a build of `zenoh_cpp` that includes all the features required by `rmw_zenoh` (see https://github.com/ros2/rmw_zenoh/blob/56f2688994600a176b9b26f7c2f23c00487e220c/zenoh_cpp_vendor/CMakeLists.txt#L18) and also all the default features enabled, so we want to force the use of the conda-forge version, instead of having `zenoh_cpp_vendor` builds its own limited variant of `zenoh`, that could conflict with the conda-forge' zenoh.
* `gz-*-vendor` packages: In conda-forge, we have a build of the gz libraries that support Python bindings (differently from vendor packages, see https://github.com/gazebo-tooling/gz_vendor/issues/2) so we want to use them. However, the satisfied check in gz-math-vendor and other packages is `EXACT` by default (see https://github.com/gazebo-release/gz_math_vendor/blob/51d2e68e872d2f071bdf40fe7017b9a00ddd8a19/CMakeLists.txt#L31-L37), and the check is not easily disabled by CMake arguments (just from environment variables, but that is tricky for example to set via colcon.pkg files).

At the moment, we handle that with a lot of patches (see for example https://github.com/RoboStack/ros-kilted/blob/main/patch/ros-kilted-zenoh-cpp-vendor.patch, https://github.com/RoboStack/ros-kilted/blob/main/patch/ros-kilted-gz-cmake-vendor.patch, https://github.com/RoboStack/ros-kilted/blob/main/patch/ros-kilted-gz-rendering-vendor.patch and all similar patch in that directory). However, patches are typically cumbersome to maintain, as you need to frequently update them if patched project changes.

To deal with this, this PR adds the `AMENT_VENDOR_POLICY` CMake option, whose documentation is included in the code, and is reported in the following:

~~~
#   AMENT_VENDOR_POLICY: String option that specifies how ament_vendor behaves,
#                        the allowed values are listed in the following.
#     DEFAULT: Vendor if ``SATISFIED`` argument is not supplied or false,
#              do not vendor otherwise.
#     FORCE_BUILD_VENDOR: Always vendor, independently of the value of the
#                   ``SATISFIED`` argument.
#     NEVER_VENDOR: Never vendor, and raise an error if ``SATISFIED`` argument
#                   is not supplied or false.
#     NEVER_VENDOR_IGNORE_SATISFIED_CHECK: Never vendor, and do not raise
#                   an error even if ``SATISFIED`` argument is not supplied
#                   or false. This option is in unsupported by most packages,
#                   so use at your own risk, as it could break the buid.
~~~

~One downside of having this option, is that now it is not anymore trivial to know if `ament_vendor` actually vendored something or not, while before it was sufficient to check if the variable passed to `SATISFIED` was `TRUE` or not (see for example https://github.com/gazebo-release/gz_sim_vendor/blob/423f5eaf34d1c69735c99704c610e434e902886e/CMakeLists.txt#L90). For this reason, the PR also adds a `IS_VENDORED_OUTPUT_VARIABLE_NAME` argument that pass the variable name that is populated by `ament_vendor` reporting if the package is actually vendored or not.~

To check if a package was actually vendored, you can check if the target name passed to `ament_vendor` exists. For example, the `gz-vendor` packages can check that with:

~~~cmake
ament_vendor(${LIB_NAME_UNDERSCORE}_vendor
  SATISFIED ${${LIB_NAME_FULL}_FOUND}
  VCS_URL https://github.com/gazebosim/${GITHUB_NAME}.git
  VCS_VERSION ${GITHUB_NAME}${LIB_VER_MAJOR}_${LIB_VER}${LIB_VER_SUFFIX}
  CMAKE_ARGS
    -DSKIP_PYBIND11:BOOL=ON
  GLOBAL_HOOK
)

# [...]

# Code that only should done if the package is actually vendored
if(TARGET ${LIB_NAME_UNDERSCORE}_vendor)
  ament_environment_hooks("${CMAKE_CURRENT_SOURCE_DIR}/${PROJECT_NAME}.dsv.in")
  # Create a dummy .sh file needed for ament_package to source the .dsv file.
  # See https://github.com/ament/ament_package/issues/145
  file(WRITE ${CMAKE_CURRENT_BINARY_DIR}/${PROJECT_NAME}.sh "# Dummy .sh file needed for .dsv file to be sourced.")
  ament_environment_hooks("${CMAKE_CURRENT_BINARY_DIR}/${PROJECT_NAME}.sh")
endif()
~~~

traversaro committed:
ament_vendor: Add IS_VENDORED_OUTPUT_VARIABLE_NAME argument and AMENT_VENDOR_POLICY CMake option

Signed-off-by: Silvio <silvio.traversaro@iit.it>
Signed-off-by: Silvio Traversaro <silvio.traversaro@gbionics.ai>

YuanYuYuan commented:
As one of the [ros-nix-overlay](https://github.com/lopsided98/nix-ros-overlay) users, I'm excited about this solution :smile:

traversaro commented:
> As one of the [ros-nix-overlay](https://github.com/lopsided98/nix-ros-overlay) users, I'm excited about this solution 😄

Great, feel free to comment if there is anything that could be improved from the nix point of view, as I a not so familiar with nix.

sloretz commented on ament_cmake_vendor_package/cmake/ament_vendor.cmake:
```diff
@@ -48,6 +48,31 @@
+#
+# Beside proper CMake macro arguments, the macro also is influenced by the
+# following CMake advanced options, that can be set from the CMake command
+# line when the project that contains the 'ament_vendor' call is configured.
```
Documentation nit so it looks like I reviewed the PR 🙃 

* passive voice -> active voice

```suggestion
# The AMENT_VENDOR_POLICY cache entry and the SATISFIED argument
# control whether or not this function builds the vendor package.
# If you want something other than the default, set AMENT_VENDOR_POLICY via the
# command line:
#
#    * If you are using cmake directly: cmake -DAMENT_VENDOR_POLICY:STRING=DEFAULT ...
#    * If you are using colcon: colcon --cmake-args -DAMENT_VENDOR_POLICY:STRING=DEFAULT ...
```

traversaro replied:
Thanks, applied in https://github.com/ament/ament_cmake/pull/592/commits/6c5844f8caedcb107a124f8c312338ba8195b400 .

sloretz commented on ament_cmake_vendor_package/cmake/ament_vendor.cmake:
```diff
@@ -48,6 +48,31 @@
+#
+# To check if a package has been actually vendored, downstream users of
+# ``ament_vendor` can pass a variable name to IS_VENDORED_OUTPUT_VARIABLE_NAME
+# argument, and check its value.
```
nit, use active voice

```suggestion
# To check if a package has been actually vendored, pass a variable name
# into the argument IS_VENDORED_OUTPUT_VARIABLE_NAME, and check
# if the variable is TRUE after you call `ament_vendor`.
```

traversaro replied:
https://github.com/ament/ament_cmake/pull/592/commits/6c5844f8caedcb107a124f8c312338ba8195b400

sloretz commented on ament_cmake_vendor_package/cmake/ament_vendor.cmake:
```diff
@@ -93,15 +118,57 @@ macro(ament_vendor TARGET_NAME)
-      message(STATUS "Forcing vendor package build for '${TARGET_NAME}', which is already satisfied")
+      message(STATUS "Forcing vendor package build for '${TARGET_NAME}', which is already satisfied as AMENT_VENDOR_POLICY is set to FORCE_BUILD_VENDOR")
     endif()
+  elseif(AMENT_VENDOR_POLICY STREQUAL "NEVER_VENDOR")
+    if(NOT _ARG_SATISFIED)
+      message(FATAL_ERROR "Error as SATISFIED argument is not TRUE, but AMENT_VENDOR_POLICY is set to NEVER_VENDOR")
+    endif()
+    set(_call_ament_vendor FALSE)
+  elseif(AMENT_VENDOR_POLICY STREQUAL "NEVER_VENDOR_IGNORE_SATISFIED_CHECK")
+    if(NOT _ARG_SATISFIED)
+      message(STATUS "Not vendoring even if SATISFIED is not TRUE as AMENT_VENDOR_POLICY is set to NEVER_VENDOR_IGNORE_SATISFIED_CHECK")
+    endif()
+    set(_call_ament_vendor FALSE)
+  else()
```
@cottsay this seems to differ from your suggestion of using the presence of the `SATISFIED` argument to enable the check in https://github.com/ament/ament_cmake/pull/552#issuecomment-2407690513 , but I think it accomplishes the same thing as `FORCE_BUILD_VENDOR` by building the package regardless of `SATISFIED` being set.

It seems like [we'd need CMake 3.31 to handle the case where SATISFIED is given but empty string](https://cmake.org/cmake/help/latest/policy/CMP0174.html#policy:CMP0174), so I think if we used `if(DEFINED...` then the warning would be skipped if SATISFIED was provided with a variable containing the empty string.

cottsay replied:
It really depends on how @traversaro thinks that an unconditionally vendored package should behave in the presence of `NEVER_VENDOR_IGNORE_SATISFIED_CHECK`. The naming makes me think that a vendored package which doesn't even check for satisfaction would NOT raise an error, in which case, I agree that using `DEFINED` to check if `_ARG_SATISFIED` was even supplied would be a better behavior than raising an error 100% of the time.

traversaro replied:
I am not sure I follow here. The value of `_ARG_SATISFIED` is normalized earlier:

~~~
  if(NOT _ARG_SATISFIED)
    set(_ARG_SATISFIED FALSE)
  endif()
~~~

after that check, it is not possible for `_ARG_SATISFIED` to be undefined or to be defined as an empty string. The only possible values after that check are `FALSE` or all values that are considered true by CMake, i.e. *1, ON, YES, TRUE, Y, or a non-zero number (including floating point numbers)*, see https://cmake.org/cmake/help/latest/command/if.html#basic-expressions .

sloretz commented on ament_cmake_vendor_package/cmake/ament_vendor.cmake:
```diff
@@ -93,15 +118,57 @@ macro(ament_vendor TARGET_NAME)
+      message(STATUS "Forcing vendor package build for '${TARGET_NAME}', which is already satisfied as AMENT_VENDOR_POLICY is set to FORCE_BUILD_VENDOR")
     endif()
+  elseif(AMENT_VENDOR_POLICY STREQUAL "NEVER_VENDOR")
+    if(NOT _ARG_SATISFIED)
+      message(FATAL_ERROR "Error as SATISFIED argument is not TRUE, but AMENT_VENDOR_POLICY is set to NEVER_VENDOR")
+    endif()
```
I assume `NEVER_VENDOR_IGNORE_SATISFIED_CHECK` is the option you intend to use in `robostack`? If so, what's the use case for `NEVER_VENDOR` on it's own?

What do you think about downgrading this from `FATAL_ERROR` to `WARNING` and removing `NEVER_VENDOR_IGNORE_SATISFIED_CHECK` to simplify the policy?

traversaro replied:
Actually I am currently using both, depending on the case.
* If it is a deps I am really familiar with, and I know the version conda will install is going to be ok --> `NEVER_VENDOR_IGNORE_SATISFIED_CHECK`. For example I do that for gz-*-vendor packages, basically implementing what they do with the `GZ_RELAX_VERSION_MATH` env variable (see https://github.com/gazebo-release/gz_math_vendor/blob/51d2e68e872d2f071bdf40fe7017b9a00ddd8a19/CMakeLists.txt#L35), but using CMake variables
* If it is a dependency that I a not familiar with, and perhaps frequently changes its API, but still I want it to be provided by conda-forge for compatibility with the broader ecosystem, I want it to still not vendor it, but give me a clear (and as early as possible) error if for any reason the "wrong" version is installed.

traversaro replied:
In general, `NEVER_VENDOR` is an option that I imagine can be considered more and less "supported", while if a user uses `NEVER_VENDOR_IGNORE_SATISFIED_CHECK` we really needs to make sure that he knows what he is doing.

sloretz reviewed (approved):
Thank you for the PR!

I have a couple nits and a question, but this code seems to do what it says.

traversaro committed:
Improve documentation by using active voice

Co-authored-by: Shane Loretz <shane.loretz@gmail.com>
Signed-off-by: Silvio Traversaro <silvio@traversaro.it>
Signed-off-by: Silvio Traversaro <silvio.traversaro@gbionics.ai>

cottsay commented on ament_cmake_vendor_package/cmake/ament_vendor.cmake:
```diff
@@ -48,6 +48,35 @@
 #   project, expose the external project globally to any downstream CMake
 #   projects.
 # :type GLOBAL_HOOK: option
+# :param IS_VENDORED_OUTPUT_VARIABLE_NAME: the name of the variable that
+#   will be set to ``TRUE`` if the package has been vendored, or ``FALSE``
+#   otherwise.
+# :type IS_VENDORED_OUTPUT_VARIABLE_NAME: string
```
Elsewhere in ROS 2 we're already using `if(TARGET target_name)` for this purpose. Is there some reason that approach isn't sufficient?

For example: https://github.com/ros2/libyaml_vendor/blob/c69b99721471d47f88042b44e510cfefafb385b9/CMakeLists.txt#L27

traversaro replied:
Good point! I did not think about this approach, and apparently I am not the only one (all gz vendor packages duplicate the satifisfied condition to check it the package was vendored or not, see https://github.com/gazebo-release/gz_math_vendor/blob/51d2e68e872d2f071bdf40fe7017b9a00ddd8a19/CMakeLists.txt#L67). But using the target defined by the ExternalProject make total sense, I will revise the PR to suggest doing that instead.

traversaro replied:
This was fixed in https://github.com/ament/ament_cmake/pull/592/commits/a935bb3d220fd5b832d8b7bbf34ae65e2425acf5 .

cottsay commented on ament_cmake_vendor_package/cmake/ament_vendor.cmake:
```diff
@@ -48,6 +48,35 @@
+#                   is not supplied or false.
+#     NEVER_VENDOR_IGNORE_SATISFIED_CHECK: Never vendor, and do not raise
+#                   an error even if ``SATISFIED`` argument is not supplied
+#                   or false. This option is in unsupported by most packages,
```
```suggestion
#                   or false. This option is unsupported by most packages,
```

cottsay commented on ament_cmake_vendor_package/cmake/ament_vendor.cmake:
```diff
@@ -93,15 +122,57 @@ macro(ament_vendor TARGET_NAME)
+  set(_AMENT_VENDOR_POLICY_DOCS "Specify how ament_vendor behaves, allowed values are DEFAULT, FORCE_BUILD_VENDOR, NEVER_VENDOR and NEVER_VENDOR_IGNORE_SATISFIED_CHECK.")
+  set(AMENT_VENDOR_POLICY "DEFAULT" CACHE STRING ${_AMENT_VENDOR_POLICY_DOCS})
+  set_property(CACHE AMENT_VENDOR_POLICY PROPERTY STRINGS "DEFAULT" "FORCE_BUILD_VENDOR" "NEVER_VENDOR" "NEVER_VENDOR_IGNORE_SATISFIED_CHECK")
+  mark_as_advanced(AMENT_VENDOR_POLICY)
```
Is there a reason to hide this?

traversaro replied:
Probably not, I typically hide options defined in functions to avoid littering the downstream projects option space (for example, last time I checked calling `find_package(rclcpp)` basically hides all project specific options if they start with a letter later in the alphabet), but in this case if `ament_vendor` is used in a project, then the main purpouse of a project is to be a vendor package, and then it make sense for the option to be visible.

traversaro committed:
Remove IS_VENDORED_OUTPUT_VARIABLE_NAME output variable

Signed-off-by: Silvio Traversaro <silvio.traversaro@gbionics.ai>

traversaro committed:
Fix typo

Co-authored-by: Scott K Logan <logans@cottsay.net>
Signed-off-by: Silvio Traversaro <silvio@traversaro.it>
Signed-off-by: Silvio Traversaro <silvio.traversaro@gbionics.ai>

traversaro committed:
Do not mark_as_advanced AMENT_VENDOR_POLICY

Signed-off-by: Silvio Traversaro <silvio.traversaro@gbionics.ai>

traversaro commented:
Thanks a lot for the review @sloretz and @cottsay, I should have addressed all your comments, and sorry for the delay in the reply. 

Downstream projects are adopting local workarounds due to this option not being available (see for example https://github.com/ros2/rmw_zenoh/pull/908) so it would be great to have this option in ament_cmake_vendor_package itself.

mjcarroll commented on ament_cmake_vendor_package/cmake/ament_vendor.cmake:
```diff
@@ -96,12 +120,45 @@ macro(ament_vendor TARGET_NAME)
+
+  if(FORCE_BUILD_VENDOR_PKG AND NOT AMENT_VENDOR_POLICY STREQUAL "FORCE_BUILD_VENDOR")
+    message(DEPRECATION "FORCE_BUILD_VENDOR_PKG set to ON detected, FORCE_BUILD_VENDOR_PKG variable is deprecated, please set AMENT_VENDOR_POLICY to FORCE_BUILD_VENDOR instead.")
+    set(CMAKE_BUILD_TYPE "FORCE_BUILD_VENDOR" CACHE STRING ${_AMENT_VENDOR_POLICY_DOCS} FORCE)
```
I don't think we want to set CMAKE_BUILD_TYPE here, right?

traversaro replied:
Good point, fixed in https://github.com/ament/ament_cmake/pull/592/changes/b95318c5bf0fb3d7ea466acabae0d9114c321712 .

mjcarroll commented on ament_cmake_vendor_package/cmake/ament_vendor.cmake:
```diff
@@ -96,12 +120,45 @@ macro(ament_vendor TARGET_NAME)
     OFF)
+  mark_as_advanced(FORCE_BUILD_VENDOR_PKG)
+
+  set(_AMENT_VENDOR_POLICY_DOCS "Specify how ament_vendor behaves, allowed values are DEFAULT, FORCE_BUILD_VENDOR, NEVER_VENDOR and NEVER_VENDOR_IGNORE_SATISFIED_CHECK.")
```
Since this is a macro, this leaks to the parent scope. Does this conflict with multiple invocations?  Do we need to unset at the end of the scope here?

traversaro replied:
Good point, fixed in https://github.com/ament/ament_cmake/pull/592/changes/b95318c5bf0fb3d7ea466acabae0d9114c321712 .

mjcarroll commented on ament_cmake_vendor_package/cmake/ament_vendor.cmake:
```diff
@@ -96,12 +120,45 @@ macro(ament_vendor TARGET_NAME)
+
+  set(_AMENT_VENDOR_POLICY_DOCS "Specify how ament_vendor behaves, allowed values are DEFAULT, FORCE_BUILD_VENDOR, NEVER_VENDOR and NEVER_VENDOR_IGNORE_SATISFIED_CHECK.")
+  set(AMENT_VENDOR_POLICY "DEFAULT" CACHE STRING ${_AMENT_VENDOR_POLICY_DOCS})
+  set_property(CACHE AMENT_VENDOR_POLICY PROPERTY STRINGS "DEFAULT" "FORCE_BUILD_VENDOR" "NEVER_VENDOR" "NEVER_VENDOR_IGNORE_SATISFIED_CHECK")
```
What happens in the case of multiple vendor packages in the same package and the CACHE variable?

traversaro replied:
If the cache entry already exists, a call to `set(AMENT_VENDOR_POLICY "DEFAULT" CACHE STRING ${_AMENT_VENDOR_POLICY_DOCS})` will not change the existing value. In a nutshell, this means that the policy of both `ament_vendor` invocations will be controlled by a single `AMENT_VENDOR_POLICY` variable (this is for example how this is used for `zenoh_cpp_vendor` in robostack). One could think of having the possibility of setting per-`ament_vendor` policy with a `AMENT_VENDOR_POLICY_${TARGET_NAME}` variable, but as I am not aware of any use case for that, I am afraid it would be not useful complexity.

traversaro committed:
Fix wrong CMAKE_BUILD_TYPE set and remove variable pollution

Signed-off-by: Silvio Traversaro <silvio.traversaro@gbionics.ai>

traversaro commented:
@mjcarroll I should have addressed your feedback. On an unrelated note, could it make sense to enable signoff on web commits for this repo or org? See for example https://docs.github.com/en/organizations/managing-organization-settings/managing-the-commit-signoff-policy-for-your-organization#managing-compulsory-commit-signoffs-for-your-organization, this is already done in gazebosim and ros orgs, and simplifies the process of making simple web-based commits without the DCO bot complaining.

traversaro commented:
@mjcarroll @sloretz @cottsay if there is anything I can do to simplify the review of this PR, feel free to ask! It would be great to reduce the amount of patches we have downstream in robostack, and this modification is quite extensively used, so it would be great to have it merged upstram at some point, thanks a lot!

mjcarroll commented:
Pulls: ament/ament_cmake#592
Gist: https://gist.githubusercontent.com/mjcarroll/e9ee0e0a05d6c5565388631ab7adb7a2/raw/bf4498cec61724cfce2e59f0774a5c201714db3a/ros2.repos
BUILD args: --packages-above-and-dependencies ament_cmake_vendor_package
TEST args: --packages-above ament_cmake_vendor_package
ROS Distro: rolling
Job: ci_launcher
ci_launcher ran: https://ci.ros2.org/job/ci_launcher/19396
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=29217)](http://ci.ros2.org/job/ci_linux/29217/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=22006)](http://ci.ros2.org/job/ci_linux-aarch64/22006/)
* Linux-rhel [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-rhel&build=9101)](http://ci.ros2.org/job/ci_linux-rhel/9101/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=28167)](http://ci.ros2.org/job/ci_windows/28167/)

Tobias-Fischer commented:
Quick reminder @mjcarroll @sloretz @cottsay

traversaro commented:
> @mjcarroll @sloretz @cottsay if there is anything I can do to simplify the review of this PR, feel free to ask! It would be great to reduce the amount of patches we have downstream in robostack, and this modification is quite extensively used, so it would be great to have it merged upstram at some point, thanks a lot!

Hello @mjcarroll @sloretz @cottsay , friendly remainder on this, having this upstream would be really useful to drop a bunch of robostack-specific patches, and remove the need for all downstream package introduce custom specific-options, as done for example in https://github.com/ros2/rmw_zenoh/pull/908 .
